### DISCUSSION TITLE
Similarities and comparison with LMQL, Guidance

### DISCUSSION THREAD

**Discussion by ogencoglu** [outlines-dev/outlines#298] [Q&A]
Thanks for the work!

A comparison table with LMQL and Guidance and similar libraries (e.g.[ like this](https://docs.lmql.ai/en/latest/python/comparison.html)) would be great. Any possibility for that?

**Comment by rlouf** [comment]
We have this figure in our paper:

![comparison-guidance-outlines (1)](https://github.com/normal-computing/outlines/assets/3885044/31b0ad12-2d2a-49a8-9398-3590574b372d)

What metrics would you find interesting?

**Reply by endolith** [reply]
A comparison table of things that one library does that the other library doesn't, so we can choose which would be best for a given application.

**Comment by ogencoglu** [comment]
Similar features from the table would be nice. For example does it support token healing, json decoding, what types of constraints are available etc.

**Comment by Mihaiii** [comment]
Does outlines choses the top token (after considering the template constraints) from logits when sampling and therefore is only doing one inference call, even if the provided template combines standard text with generated text multiple times?

If I'm not mistaken, guidance will make a new inference call each time it hits a generative command within the template.

**Comment by wenxueru** [comment]
Does outlines support the token healing mentioned in Guidance?

**Reply by rlouf** [reply]
Soon :)

**Comment by freckletonj** [comment]
Here's my short summary from playing with these all recently:

* `lmql`: unusable morass of bugs
* `guidance`: best dev UI of all, truly great, but abandonware now
* `outlines`: no way to have a huge prompt with generations happening distributed throughout, and then named in a dictionary key to pull out later (like `guidance`)

**Reply by xiaohk** [reply]
You probably have seen this. `guidance` is now ["reborn"](https://github.com/guidance-ai/guidance/discussions/429) 😆

**Comment by brandonwillard** [comment]
> * `outlines`: no way to have a huge prompt with generations happening distributed throughout, and then named in a dictionary key to pull out later (like `guidance`)

Sounds like a good feature request!  If you don't mind, create an issue and we can determine how it can meet these/your needs.  We have a lot of big improvements and features in the pipeline, so it would be good to map this all out sooner than later.

**Comment by ogencoglu** [comment]
@freckletonj  Would you like to elaborate on your `lmql` experience here or maybe [over here](https://github.com/eth-sri/lmql/discussions/224) ?

**Comment by brandonwillard** [comment]
> For example does it support token healing, json decoding, what types of constraints are available etc.

Aside from the prompt construction features, `outlines` provides a very broad and run-time optimal approach to regex-guided generation, which also happens to cover JSON decoding.  It also doesn't require token healing.  We have a simple `pydantic` interface to the JSON decoding, too.

Presently, the library provides tools for the even broader task of (deterministic) CFG-guided generation and we plan to roll out a much more convenient interface for that feature soon.  We have a highly efficient approach to this as well.  See [our technical paper](https://arxiv.org/abs/2307.09702) for details.

More generally, `outlines` isn't trying to provide a single languages/DSL for the whole of constrained generation and/or prompting, and that might be one notable difference with `lmql`.  We're currently focusing on composable, efficient core tools and features, which are then exposed though hackable integrations with established DSLs (e.g. `pydantic` for constraints, `jinja2` for prompts) and some simple custom classes (e.g. `Sequence` for sampling).  We have a lot of core design changes in the pipeline, but this mostly sums up the overall intent of what we've done and are doing.
